lib/galaxy/jobs/runners/cli.py

```diff
@@ -177,6 +177,10 @@ def check_watched_items(self):
                 if not state == model.Job.states.OK:
                     # No need to change_state when the state is OK, this will be handled by `self.finish_job`
                     ajs.job_wrapper.change_state(state)
+                if state == model.Job.states.ERROR:
+                    # Try to find out the reason for exiting
+                    self.__handle_out_of_memory(ajs, external_job_id)
+                    self.mark_as_failed(ajs)
```

Comment:
```suggestion
                    self.work_queue.put((self.mark_as_failed, ajs))
```

Reply:
Thanks, happy to go with this if you say it is more correct. I will test it locally to make sure it works and then commit the suggestion.